pkg/scheduler/algorithm/priorities/metadata.go

```diff
@@ -123,32 +123,61 @@ func getSelector(pod *v1.Pod, sl corelisters.ServiceLister, cl corelisters.Repli
 		}
 	}
 
-	if rcs, err := cl.GetPodControllers(pod); err == nil {
-		for _, rc := range rcs {
-			labelSet = labels.Merge(labelSet, rc.Spec.Selector)
+	rcPod, rsPod, stsPod := false, false, false
+	for _, ref := range pod.OwnerReferences {
+		if ref.Controller != nil && *ref.Controller == true {
+			switch ref.Kind {
+			case "ReplicationController":
+				if ref.APIVersion == "v1" {
+					rcPod = true
+				}
+			case "ReplicaSet":
+				if ref.APIVersion == "apps/v1" || ref.APIVersion == "apps/v1beta2" {
+					rsPod = true
+				}
+			case "StatefulSet":
+				if ref.APIVersion == "apps/v1" || ref.APIVersion == "apps/v1beta1" || ref.APIVersion == "apps/v1beta2" {
+					stsPod = true
+				}
+			default:
+				continue
+			}
+			break
+		}
+	}
+
+	if rcPod {
+		if rcs, err := cl.GetPodControllers(pod); err == nil {
```

Comment:
@liggitt since there is only one controller managing a pod, why do we need to list all controllers in the pod's namespace and match them against the pod labels? the OwnerReference contains the controller name, so we can just query that specific controller and get its labels, right? I get that this has slightly difference semantics than the current implementation because we could have multiple with overlapping selectors, but this isn't a case that this priority was meant to handle.

Reply:
looking up the actual controlling object makes sense to me

Reply:
Any reason not to do the same for ReplicaSets and StatefulSets as well?